docs/content/examples/tutorials/mailserver-behind-proxy.md

```diff
@@ -2,126 +2,410 @@
+    smtp/inet/postscreen_upstream_proxy_protocol=haproxy
+    submission/inet/smtpd_upstream_proxy_protocol=haproxy
+    submissions/inet/smtpd_upstream_proxy_protocol=haproxy
     ```
 
-    Keep in mind that it is necessary to use port `10993` here. More information below at `dovecot` configuration.
+    [`postscreen_upstream_proxy_protocol`][postfix-docs::settings::postscreen_upstream_proxy_protocol] and [`smtpd_upstream_proxy_protocol`][postfix-docs::settings::smtpd_upstream_proxy_protocol] both specify the protocol type used by a proxy. `haproxy` represents the PROXY protocol.
+
+    ---
+
+    Dovecot via [`dovecot.cf`][docs::overrides::dovecot]:
+
+    ```cf  title="docker-data/dms/config/dovecot.cf"
+    haproxy_trusted_networks = 172.16.42.2
+
+    service imap-login {
+      inet_listener imap {
+        haproxy = yes
+      }
+
+      inet_listener imaps {
+        haproxy = yes
+      }
+    }
+
+    service pop3-login {
+      inet_listener pop3 {
+        haproxy = yes
+      }
+
+      inet_listener pop3s {
+        haproxy = yes
+      }
+    }
+
+    service managesieve-login {
+      inet_listener sieve {
+        haproxy = yes
+      }
+    }
+    ```
+
+    - [`haproxy_trusted_networks`][dovecot-docs::settings::haproxy-trusted-networks] must reference the reverse proxy IP, or a wider subnet using CIDR notation.
+    - [`haproxy = yes`][dovecot-docs::service-config::haproxy] for the TCP listeners of each login service.
+
+!!! warning "Internal traffic (_within the network or DMS itself_)"
+
+    - Direct connections to DMS from other containers within the internal network will be rejected when they don't provide the required PROXY header.
+    - This can also affect services running within the DMS container itself if they attempt to make a connection and aren't PROXY protocol capable.
+
+    ---
+
+    A solution is to configure alternative service ports that offer PROXY protocol support (as shown next).
+
+    Alternatively routing connections to DMS through the local reverse proxy via [DNS query rewriting][gh-dms::dns-rewrite-example] can work too.
+
+??? example "Configuring services with separate ports for PROXY protocol"
+
+    In this example we'll take the original service ports and add `10000` for the new PROXY protocol service ports.
+
+    Traefik labels will need to update their service ports accordingly (eg: `.loadbalancer.server.port=10465`).
+
+    ---
+
+    Postfix config now requires [our `user-patches.sh` support][docs::overrides::user-patches] to add new services in `/etc/postfix/master.cf`:
+
+    ```bash title="docker-data/dms/config/user-patches.sh"
+    #!/bin/bash
+
+    # Duplicate the config for the submission(s) service ports (587 / 465) with adjustments for the PROXY ports (10587 / 10465) and `syslog_name` setting:
+    postconf -Mf submission/inet | sed -e s/^submission/10587/ -e 's/submission/submission-proxyprotocol/' >> /etc/postfix/master.cf
```

Comment:
I assume you run this inside the container? Wouldn't it be lost when the container is restarted?

Reply:
> I assume you run this inside the container?

Yes, as the docs mention prior, this is `user-patches.sh`. I've annotated the codefence and provided a link reference to the relevant docs page, so the reader should already have this context.

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/d719d323-cdcf-4cfd-8da8-b709ac20faad)

> Wouldn't it be lost when the container is restarted?

It would be lost when the container is recreated, which is what we advise for "restarting" (destroy the previous instance).

Otherwise it'd be appended each time since the container persists internal state regardless of a volume mount when it's restarted (_or a more manual stop/start_). Which is pretty bad, we have several issues open about such bugs.

Ideally we'd have our internal base config stored elsewhere to be copied to the proper locations at each restart so that we can avoid that issue. It'll be a breaking change though as some users volume mount configs directly to the internal location and some of those are files we may modify, addressing the issue may wipe their direct mounted volume content (_one would hope they have a backup, or at least read breaking changes in our changelog_).

Reply:
Oh I see, missed it was part of user-patches.sh